runtime/participant-runtime.ss

```diff
@@ -767,3 +768,379 @@
+    ('stdio (.call IoChannel .make)) ; TODO: Initialize io:context object here
+    ('libp2p (let ()
+     (def my-nickname (hash-get options 'my-nickname))
+     (def host-address (hash-get options 'host-address))
+     (def dest-address (hash-get options 'dest-address))
+     (.call Libp2pChannel .make my-nickname host-address dest-address)))
+    (else (error "Invalid off-chain channel selection"))))
+
+
+;; TODO: Eventually upstream changes to gerbil-libp2p to accept passing
+;; a seckey via an environment variable, instead of a file.
+
+
+(def (lookup-contact nickname: nickname contacts: contacts)
+  (for-each (lambda (contact) (displayln (.@ contact name))) contacts)
+  (find (lambda (contact) (equal? (.@ contact name) nickname)) contacts))
+
+;; ------------------ Cleanup off-chain channels
+
+
+(def (close-off-chain-channel off-chain-channel)
+  (match (.@ off-chain-channel tag)
+    ('stdio #f)
+    ('libp2p (close-libp2p-channel off-chain-channel))
+    (else (error "Invalid channel")))) ; TODO: This is an internal error,
+                                        ; ensure this is handled at cli options parsing step.
+
+;; Gracefully close the channel
+;; does libp2p-close suffice??
+(def (close-libp2p-channel chan)
+  (stop-libp2p-daemon!))
+
+;; ------------------ Send Contract Agreement
+
+
+;; FIXME: Take in participant addresses as a parameter,
+;; update this when there's support for mapping users to peerIDs via the contacts store.
+(def (send-contract-agreement agreement off-chain-channel)
+  (displayln MAGENTA "Sending Agreement to other participants...")
+  (match (.@ off-chain-channel tag)
+    ('stdio (send-contract-agreement/stdout agreement))
+    ('libp2p (let ()
+     (def libp2p-client (.@ off-chain-channel libp2p-client))
+     (def dest-address (.@ off-chain-channel dest-address))
+     (send-contract-agreement/libp2p agreement libp2p-client dest-address)))
+    (else (error "Invalid channel")))) ; TODO: This is an internal error,
+                                       ; ensure this is handled at cli options parsing step.
+
+(def (send-contract-agreement/stdout agreement)
+  (displayln MAGENTA "One line command for other participants to generate the same agreement:" END)
+  (def agreement-string (string<-json (json<- InteractionAgreement agreement)))
+  (def escaped-agreement-string (escape-json-string agreement-string))
+  (def full-cmd-string (string-append "glow start-interaction --agreement " escaped-agreement-string))
+  (displayln full-cmd-string)
+  (force-output))
+
+(def (send-contract-agreement/libp2p agreement libp2p-client dest-address)
+  (displayln MAGENTA "Sending agreement to multiaddr..." END)
+  (def agreement-string (string<-json (json<- InteractionAgreement agreement)))
+  (dial-and-send-contents libp2p-client dest-address agreement-string)
+  (displayln)
+  (force-output))
+
+;; TODO: Generalize & Upstream to gerbil-utils/json
+(def (escape-json-string json-string)
```

Comment:
Renamed this function to be more specific, and added your remarks inline as a TODO, before this can generalized as a way to shell-escape json strings.